api/v1alpha1/common.go

```diff
@@ -72,3 +77,109 @@ func UpdateCondWithFixedLen(l int, conds *ConditionedStatus, cond Condition) {
 	}
 	conds.Conditions = append(conds.Conditions, cond)
 }
+
+type Filter struct {
+	Name, Version string
+	Deprecated    bool
+}
+
+func filterMatchDeprecation(cond FilterCond, filter Filter) bool {
+	return !cond.Deprecated && filter.Deprecated
+}
+func filterMatchVersion(cond FilterCond, filter Filter) bool {
+	for _, v := range cond.Versions {
+		if filter.Version == v {
+			return true
+		}
+	}
+	return false
+}
+
+func filterMatchRegexp(cond FilterCond, filter Filter) bool {
+	if len(cond.Regexp) > 0 {
+		reg, err := regexp.Compile(cond.Regexp)
+		if err == nil && reg.MatchString(filter.Version) {
+			return true
+		}
+	}
+	return false
+}
+
+func filterMatchVersionConstraint(cond FilterCond, filter Filter) bool {
+	if len(cond.VersionConstraint) > 0 {
+		constraint, err := semver.NewConstraint(cond.VersionConstraint)
+		if err != nil {
+			return false
+		}
+		v, err := semver.NewVersion(filter.Version)
+		if err != nil {
+			return false
+		}
+		return constraint.Check(v)
+	}
+	return false
+}
+
+func Match(filterCond map[string]FilterCond, filter Filter) bool {
+	if len(filterCond) == 0 {
+		return true
+	}
+
+	cond, ok := filterCond[filter.Name]
+	if !ok {
+		return true
+	}
+
+	if !filterMatchDeprecation(cond, filter) {
+		return false
+	}
+	if filterMatchVersion(cond, filter) {
+		return true
+	}
+	if filterMatchRegexp(cond, filter) {
+		return true
+	}
+	if filterMatchVersionConstraint(cond, filter) {
+		return true
+	}
+	return false
+}
+
+func IsCondSame(c1, c2 FilterCond) bool {
+	c1Map := make(map[string]int)
+	c2Map := make(map[string]int)
+	for _, v := range c1.Versions {
```

Comment:
@bjwswang @Abirdcfly Is it reasonable to implement this to determine if the contents of an array are consistent?

Reply:
```golang
func stringSlicesEqual(s1, s2 []string) bool {
    if len(s1) != len(s2) {
        return false
    }
    set1 := make(map[string]bool)
    for _, v := range s1 {
        set1[v] = true
    }
    for _, v := range s2 {
        if !set1[v] {
            return false
        }
    }
    return true
}

```

Reply:
just for versions compare.
```go
// "k8s.io/apimachinery/pkg/util/sets"
sets.NewString(c1.Versions).Equal(sets.NewString(c2.Versions))
```

Reply:
/wtd verify two string slice equals or not

Reply:
> ```go
> func stringSlicesEqual(s1, s2 []string) bool {
>     if len(s1) != len(s2) {
>         return false
>     }
>     set1 := make(map[string]bool)
>     for _, v := range s1 {
>         set1[v] = true
>     }
>     for _, v := range s2 {
>         if !set1[v] {
>             return false
>         }
>     }
>     return true
> }
> ```

version1 = [v1, v1, v2]
version2 = [v2, v2, v2]
should return false, but.